test/unit/auth/tenant-manager.spec.ts

```diff
@@ -0,0 +1,590 @@
+import {Tenant, TenantOptions, TenantServerResponse, ListTenantsResult} from '../../../src/auth/tenant';
+import {TenantManager} from '../../../src/auth/tenant-manager';
+import {AuthClientErrorCode, FirebaseAuthError} from '../../../src/utils/error';
+
+chai.should();
+chai.use(sinonChai);
+chai.use(chaiAsPromised);
+
+const expect = chai.expect;
+
+describe('TenantManager', () => {
+  const TENANT_ID = 'tenantId';
+  let mockApp: FirebaseApp;
+  let tenantManager: TenantManager;
+  let nullAccessTokenTenantManager: TenantManager;
+  let malformedAccessTokenTenantManager: TenantManager;
+  let rejectedPromiseAccessTokenTenantManager: TenantManager;
+
+
+  beforeEach(() => {
+    mockApp = mocks.app();
+    tenantManager = new TenantManager(mockApp);
+    nullAccessTokenTenantManager = new TenantManager(
+        mocks.appReturningNullAccessToken());
+    malformedAccessTokenTenantManager = new TenantManager(
+        mocks.appReturningMalformedAccessToken());
+    rejectedPromiseAccessTokenTenantManager = new TenantManager(
+        mocks.appRejectedWhileFetchingAccessToken());
+
+  });
+
+  afterEach(() => {
+    return mockApp.delete();
+  });
+
+  describe('authForTenant()', () => {
+    const invalidTenantIds = [null, NaN, 0, 1, true, false, '', [], [1, 'a'], {}, { a: 1 }, _.noop];
+    invalidTenantIds.forEach((invalidTenantId) => {
+      it('should throw given invalid tenant ID: ' + JSON.stringify(invalidTenantId), () => {
+        expect(() => {
+          return tenantManager.authForTenant(invalidTenantId as any);
+        }).to.throw('The tenant ID must be a valid non-empty string.');
+      });
+    });
+
+    it('should return a TenantAwareAuth with the expected tenant ID', () => {
+      expect(tenantManager.authForTenant(TENANT_ID).tenantId).to.equal(TENANT_ID);
+    });
+
+    it('should return a TenantAwareAuth with read-only tenant ID', () => {
+      expect(() => {
+        (tenantManager.authForTenant(TENANT_ID) as any).tenantId = 'OTHER_TENANT_ID';
+      }).to.throw('Cannot assign to read only property \'tenantId\' of object \'#<TenantAwareAuth>\'');
+    });
+
+    it('should cache the returned TenantAwareAuth', () => {
+      const tenantAwareAuth1 = tenantManager.authForTenant('tenantId1');
+      const tenantAwareAuth2 = tenantManager.authForTenant('tenantId2');
+      expect(tenantManager.authForTenant('tenantId1')).to.equal(tenantAwareAuth1);
+      expect(tenantManager.authForTenant('tenantId2')).to.equal(tenantAwareAuth2);
+      expect(tenantAwareAuth1).to.not.be.equal(tenantAwareAuth2);
+      expect(tenantAwareAuth1.tenantId).to.equal('tenantId1');
+      expect(tenantAwareAuth2.tenantId).to.equal('tenantId2');
+    });
+  });
+
+  describe('getTenant()', () => {
+    const tenantId = 'tenant_id';
+    const serverResponse: TenantServerResponse = {
+      name: 'projects/project_id/tenants/tenant_id',
+      displayName: 'TENANT_DISPLAY_NAME',
```

Comment:
Use only letters, digits and hyphens to to be consistent with requirements

Reply:
Done.